spring-cloud-stream-binder-rabbit-core/src/main/java/org/springframework/cloud/stream/binder/rabbit/provisioning/RabbitExchangeQueueProvisioner.java

```diff
@@ -0,0 +1,490 @@
+		if (properties.getDeadLetterExchange() == null) {
+			return properties.getPrefix() + RabbitCommonProperties.DEAD_LETTER_EXCHANGE;
+		}
+		else {
+			return properties.getDeadLetterExchange();
+		}
+	}
+
+
+	private void declareQueue(String beanName, Queue queue) {
+		try {
+			this.rabbitAdmin.declareQueue(queue);
+		}
+		catch (AmqpConnectException e) {
+			if (this.logger.isDebugEnabled()) {
+				this.logger.debug("Declaration of queue: " + queue.getName() + " deferred - connection not available");
+			}
+		}
+		addToAutoDeclareContext(beanName, queue);
+	}
+
+	private Map<String, Object> queueArgs(String queueName, RabbitCommonProperties properties, boolean isDlq) {
+		Map<String, Object> args = new HashMap<>();
+		if (!isDlq) {
+			if (properties.isAutoBindDlq()) {
+				String dlx;
+				if (properties.getDeadLetterExchange() != null) {
+					dlx = properties.getDeadLetterExchange();
+				}
+				else {
+					dlx = applyPrefix(properties.getPrefix(), "DLX");
+				}
+				args.put("x-dead-letter-exchange", dlx);
+				String dlRk;
+				if (properties.getDeadLetterRoutingKey() != null) {
+					dlRk = properties.getDeadLetterRoutingKey();
+				}
+				else {
+					dlRk = queueName;
+				}
+				args.put("x-dead-letter-routing-key", dlRk);
+			}
+			additionalArgs(args, properties.getExpires(), properties.getMaxLength(), properties.getMaxLengthBytes(),
+					properties.getMaxPriority(), properties.getTtl());
+		}
+		else {
+			if (properties.getDlqDeadLetterExchange() != null) {
+				args.put("x-dead-letter-exchange", properties.getDlqDeadLetterExchange());
+			}
+			if (properties.getDlqDeadLetterRoutingKey() != null) {
+				args.put("x-dead-letter-routing-key", properties.getDlqDeadLetterRoutingKey());
+			}
+			additionalArgs(args, properties.getDlqExpires(), properties.getDlqMaxLength(),
+					properties.getDlqMaxLengthBytes(), properties.getDlqMaxPriority(), properties.getDlqTtl());
+		}
+		return args;
+	}
+
+	private void additionalArgs(Map<String, Object> args, Integer expires, Integer maxLength, Integer maxLengthBytes,
+								Integer maxPriority, Integer ttl) {
+		if (expires != null) {
+			args.put("x-expires", expires);
+		}
+		if (maxLength != null) {
+			args.put("x-max-length", maxLength);
+		}
+		if (maxLengthBytes != null) {
+			args.put("x-max-length-bytes", maxLengthBytes);
+		}
+		if (maxPriority != null) {
+			args.put("x-max-priority", maxPriority);
+		}
+		if (ttl != null) {
+			args.put("x-message-ttl", ttl);
+		}
+	}
+
+
+	public static String applyPrefix(String prefix, String name) {
+		return prefix + name;
+	}
+
+	private Exchange buildExchange(RabbitCommonProperties properties, String exchangeName) {
+		try {
+			// TODO Make the ctor public in Spring-AMQP - AMQP-695
+			Constructor<ExchangeBuilder> ctor = ExchangeBuilder.class.getDeclaredConstructor(String.class, String.class);
+			ReflectionUtils.makeAccessible(ctor);
+			ExchangeBuilder builder = ctor.newInstance(exchangeName, properties.getExchangeType());
```

Comment:
I think we can get rid of this reflection hack already, since we refactor the code anyway.